src/Starcounter.Database.Extensions/PostCommitTransactor.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using Starcounter.Database.ChangeTracking;
+
+namespace Starcounter.Database.Extensions
+{
+    public class PostCommitTransactor : TransactorBase
+    {
+        readonly AsyncLocal<List<KeyValuePair<Type, Change>>> _lastChanges = new AsyncLocal<List<KeyValuePair<Type, Change>>>();
```

Comment:
I think we should strive to keep a transactor state as minimal as possible to simplify things and avoid reentrancy issues.
We have a context for that. My suggestion are to 
1. add forth and logically last context-related callback: before we enter inner transaction. This callback would be the first one and it will create a custom context
2. pass a context to `LeftContext` delegate.
3. probably rename callbacks as current names don't convey the idea of when they are called and thus one need to check it with the code.

Reply:
If I understood the idea correctly, then it's implemented via custom `ITransactorContext` where each transactor decorator can save whatever is needed there.

Reply:
If I understood the idea correctly, then it's implemented via custom `ITransactorContext` where each transactor decorator can save whatever is needed there.

Reply:
Basically it, but don't you think having two contexts is confusing? I wonder if we could have a single context utilizing two-phase initialization here - create ContextBase in the `CreateTransactorContext` callback and then initialize it with real IDatabaseContext in EnterContext. WDYT?

Reply:
I would prefer to avoid such usage, because `IDatabaseContext` is only valid within a database transaction, while `ITransactorContext` is valid before and after the transaction.

Reply:
I respect this point. Having them separate is cleaner with respect to type safety. I just wanted to highlight the fact that these two contexts always go hand by hand. And TransactorContext is never leaked to a user code so nobody could mess up with it and use it as IDatabaseContext at inappropriate time.
But go ahead if you're comfortable with two classes.

Reply:
There methods are `protected`, means user code can easily implement a custom decorator with access to `ITransactorContext` and attempt to abuse `IDatabaseContext`, thus I would prefer to keep two separate classes.

Reply:
>means user code can easily implement a custom decorator with access to ITransactorContext and attempt to abuse IDatabaseContext

WDYM? in what way?

Reply:
```cs
public class CustomPostCommitTransactor : PostCommitTransactor
{
    protected override void LeftContext(PostCommitTransactorContext transactorContext)
    {
        // DatabaseContext is no longer valid here and can be misused.
        var db = transactorContext.DatabaseContext;
    }
}
```

Reply:
Ah, ofc. I just consider writing decorators as an advanced task that requires more thoughtful coding anyway.
(And I thought about TransactorContext actually implementing IDatabaeContext instead of having it as a property. Though it changes nothing for the safety discussion we are having.)

Reply:
I have implemented a pair of:

- `EnterDatabaseContext` & `LeaveDatabaseContext`.
- `EnterTransactorContext` & `LeaveTransactorContext`.

And now looking at the code .. thinking that it's better to sacrifice some safety for the sake of code niceness.

Reply:
Actually, it's not that bad as I first thought. Having single context would still require four different method, something like:

0. `CreateTransactorContext`.
1. `EnterDatabaseContext`.
2. `LeaveDatabaseContext`.
3. `DisposeTransactorContext`.

Which makes it even uglier, and not safe. In my opinion.

Reply:
Yes, four methods are a must as far as I'm concerned. My concern was that passing around two contexts is confusing. If you're fine with that - it's ok.